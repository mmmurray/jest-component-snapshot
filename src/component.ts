import crypto from 'crypto'
import { unlink, writeFile } from 'fs-extra'
import { join } from 'path'
import { loadPage } from './puppeteer'
import ComponentElement from './types/component-element'

const getHtml = (element: ComponentElement) => {
  let elementHtml = element
  const styles: string[] = []

  return `<!doctype html><html><head><style>${styles.join(
    '\n',
  )}</style></head><body>${elementHtml}</body></html>`
}

const loadComponentIntoPage = async (element: ComponentElement) => {
  const { page, teardown: teardownPage } = await loadPage()

  const html = getHtml(element)
  const id = crypto.randomBytes(20).toString('hex')
  const filePath = join(__dirname, `test-${id}.html`)
  await writeFile(filePath, html)

  await page.goto(`file://${filePath}`)

  const teardown = async () => Promise.all([unlink(filePath), teardownPage()])

  return { page, teardown }
}

export { loadComponentIntoPage }
