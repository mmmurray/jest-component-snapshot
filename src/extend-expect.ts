import { toMatchImageSnapshot as toMatchImageSnapshotBase } from 'jest-image-snapshot'
import { toMatchSnapshot } from 'jest-snapshot'
import { a11ySnapshot, domSnapshot, imageSnapshot } from '.'

const extendExpect = () => {
  expect.extend({
    async toMatchA11ySnapshot(received) {
      return toMatchSnapshot.call(
        this,
        await a11ySnapshot(received),
        'toMatchA11ySnapshot',
      )
    },
  })

  expect.extend({
    async toMatchDomSnapshot(received) {
      return toMatchSnapshot.call(
        this,
        await domSnapshot(received),
        'toMatchDomSnapshot',
      )
    },
  })

  expect.extend({
    async toMatchImageSnapshot(received) {
      return toMatchImageSnapshotBase.call(
        this,
        await imageSnapshot(received),
        'toMatchImageSnapshot',
      )
    },
  })
}

export { extendExpect }
