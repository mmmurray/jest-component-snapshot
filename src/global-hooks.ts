import puppeteer from 'puppeteer'

const globalSetup = async () => {
  ;(global as any).browser = await puppeteer.launch({
    executablePath: process.env.CHROME_EXECUTABLE_PATH,
  })
  process.env.PUPPETEER_WS_ENDPOINT = (global as any).browser.wsEndpoint()
}

const globalTeardown = async () => {
  await (global as any).browser.close()
  ;(global as any).browser = null
}

export { globalSetup, globalTeardown }
