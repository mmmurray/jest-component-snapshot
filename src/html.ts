import ComponentElement from './types/component-element'

const getReact = () => {
  try {
    return {
      React: require('react'),
      ReactDOMServer: require('react-dom/server'),
    }
  } catch {
    return {}
  }
}

const getElementHtml = (element: ComponentElement): string => {
  if (typeof element === 'string') {
    return element
  }

  if (element instanceof HTMLElement) {
    return element.outerHTML
  }

  const { React, ReactDOMServer } = getReact()

  if (React && ReactDOMServer && React.isValidElement(element)) {
    return ReactDOMServer.renderToStaticMarkup(element)
  }
}

const getPageHtml = (element: ComponentElement) => {
  const elementHtml = getElementHtml(element)
  const styles: string[] = []

  return `<!doctype html><html><head><style>${styles.join(
    '\n',
  )}</style></head><body>${elementHtml}</body></html>`
}

export { getElementHtml, getPageHtml }
