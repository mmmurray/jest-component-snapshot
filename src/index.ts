export * from './a11y-snapshot'
export * from './dom-snapshot'
export * from './extend-expect'
export * from './global-hooks'
export * from './image-snapshot'
