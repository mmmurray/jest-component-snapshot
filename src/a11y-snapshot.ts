import { loadComponentIntoPage } from './component'
import ComponentElement from './types/component-element'

const a11ySnapshot = async (element: ComponentElement) => {
  const { page, teardown } = await loadComponentIntoPage(element)

  const snapshot = await page.accessibility.snapshot({
    interestingOnly: false,
  })

  await teardown()

  return snapshot
}

export { a11ySnapshot }
