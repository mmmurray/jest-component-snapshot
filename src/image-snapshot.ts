import { loadComponentIntoPage } from './component'
import ComponentElement from './types/component-element'

const imageSnapshot = async (element: ComponentElement) => {
  const { page, teardown } = await loadComponentIntoPage(element)

  const snapshot = await page.screenshot()

  await teardown()

  return snapshot
}

export { imageSnapshot }
