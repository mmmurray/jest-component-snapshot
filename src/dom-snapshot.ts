import pretty from 'pretty'
import { getElementHtml } from './html'
import ComponentElement from './types/component-element'

const domSnapshot = (element: ComponentElement) => {
  const html = getElementHtml(element)

  return pretty(html)
}

export { domSnapshot }
