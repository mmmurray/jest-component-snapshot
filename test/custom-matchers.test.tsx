import styled from '@emotion/styled'
import React from 'react'
import { extendExpect } from '../src'
import 'jest-component-snapshot'

extendExpect()

const stringToDomElement = html => {
  const div = document.createElement('div')
  div.innerHTML = html

  return div.firstChild
}

const StyledHeader = styled.header`
  padding: 12px;
  border-radius: 4px;
  background-color: lime;
  font-family: sans-serif;
`

const htmlStringElement = '<header><h1>Hello</h1><h2>World</h2></header>'
const domElement = stringToDomElement(htmlStringElement)
const reactElement = (
  <StyledHeader>
    <h1>Hello</h1>
    <h2>World</h2>
  </StyledHeader>
)

test('can use toMatchA11ySnapshot with HTML string', async () => {
  await expect(htmlStringElement).toMatchA11ySnapshot()
})

test('can use toMatchA11ySnapshot with DOM element', async () => {
  await expect(domElement).toMatchA11ySnapshot()
})

test('can use toMatchA11ySnapshot with React element', async () => {
  await expect(reactElement).toMatchA11ySnapshot()
})

test('can use toMatchDomSnapshot with HTML string', async () => {
  await expect(htmlStringElement).toMatchDomSnapshot()
})

test('can use toMatchDomSnapshot with DOM element', async () => {
  await expect(domElement).toMatchDomSnapshot()
})

test('can use toMatchDomSnapshot with React element', async () => {
  await expect(reactElement).toMatchDomSnapshot()
})

test('can use toMatchImageSnapshot with HTML string', async () => {
  await expect(htmlStringElement).toMatchImageSnapshot()
})

test('can use toMatchImageSnapshot with DOM element', async () => {
  await expect(domElement).toMatchImageSnapshot()
})

test('can use toMatchImageSnapshot with React element', async () => {
  await expect(reactElement).toMatchImageSnapshot()
})
