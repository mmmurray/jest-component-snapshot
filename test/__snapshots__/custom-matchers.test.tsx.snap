// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can use toMatchA11ySnapshot with DOM element: toMatchA11ySnapshot 1`] = `
"role: WebArea
children:
  -
    role: banner
    children:
      -
        role: heading
        name: Hello
        level: 1
        children:
          -
            role: text
            name: Hello
      -
        role: heading
        name: World
        level: 2
        children:
          -
            role: text
            name: World
"
`;

exports[`can use toMatchA11ySnapshot with HTML string: toMatchA11ySnapshot 1`] = `
"role: WebArea
children:
  -
    role: banner
    children:
      -
        role: heading
        name: Hello
        level: 1
        children:
          -
            role: text
            name: Hello
      -
        role: heading
        name: World
        level: 2
        children:
          -
            role: text
            name: World
"
`;

exports[`can use toMatchA11ySnapshot with React element: toMatchA11ySnapshot 1`] = `
"role: WebArea
children:
  -
    role: banner
    children:
      -
        role: heading
        name: Hello
        level: 1
        children:
          -
            role: text
            name: Hello
      -
        role: heading
        name: World
        level: 2
        children:
          -
            role: text
            name: World
"
`;

exports[`can use toMatchDomSnapshot with DOM element: toMatchDomSnapshot 1`] = `
"<header>
  <h1>Hello</h1>
  <h2>World</h2>
</header>"
`;

exports[`can use toMatchDomSnapshot with HTML string: toMatchDomSnapshot 1`] = `
"<header>
  <h1>Hello</h1>
  <h2>World</h2>
</header>"
`;

exports[`can use toMatchDomSnapshot with React element: toMatchDomSnapshot 1`] = `
"<header class=\\"css-jqpq4b\\">
  <h1>Hello</h1>
  <h2>World</h2>
</header>"
`;
